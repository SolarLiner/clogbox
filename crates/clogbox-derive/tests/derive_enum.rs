use clogbox_core::r#enum::{az::CastFrom, enum_iter, Enum, Sequential};
use clogbox_derive::Enum;
use typenum::{Unsigned, U3};

#[derive(Debug, Copy, Clone, PartialOrd, Ord, PartialEq, Eq, Enum)]
enum Inner {
    A,
    B,
    C,
    D,
}

#[derive(Debug, Copy, Clone, PartialOrd, Ord, PartialEq, Eq, Enum)]
enum Outer {
    First,
    Second(Sequential::<U3>),
    Third(Inner),
}

#[test]
fn test_inner_has_correct_size() {
    assert_eq!(4, <Inner as Enum>::Count::USIZE);
}

#[test]
fn test_outer_has_correct_size() {
    assert_eq!(8, <Outer as Enum>::Count::USIZE);
}

#[test]
fn test_inner_cast_from() {
    let actual = [0, 1, 2, 3].map(|i| Inner::cast_from(i));
    let expected = [Inner::A, Inner::B, Inner::C, Inner::D];
    assert_eq!(expected, actual);
}
#[test]
fn test_outer_enum_iter() {
    let expected = enum_iter::<Outer>()
        .map(|e| e.name().to_string())
        .collect::<Vec<_>>();
    insta::assert_csv_snapshot!(expected);
}
