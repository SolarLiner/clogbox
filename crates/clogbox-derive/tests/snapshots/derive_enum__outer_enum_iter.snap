---
source: crates/clogbox-derive/tests/derive_enum.rs
expression: expected
---
First
1
2
3
A
B
C
D
